# Home work 19.01.2025
## Dauren Abilev

Мы будем работать с набором данных Auto MPG. Это реальные исторические данные по автомобилям. Представьте ситуацию:



Вы решили купить себе ретро-тачку, например, Ford Mustang 70-го года. Нашли классный каталог в интернете. Вы смотрите характеристики, и там написано: Расход 15 mpg. Это много или мало? Мы привыкли к литрам. Нам нужно знать, сколько литров уйдет на 100 км, чтобы прикинуть бюджет.



Ваша задача:

1. Скачать этот каталог (датасет прикреплен).

2. Переведите мили на галлон (американские единицы mgp) в километры на литр. Добавьте колонку kpl.

3. В Европе и СНГ привыкли мерить расход по другому. Перевести mgp в знакомые литры на 100 км. Добавьте колонку L_100km.

4. Скажите какие машины выгоднее всего брать, чтобы меньше тратить на бензин?

In [ ]:
import pandas as pd

df = pd.read_csv('./auto-mpg.csv')

df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model year', 'origin', 'name']
print(df.head(3))
print('\n'

    mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  




miles -> km

galons -> l


## $\frac{(x) * miles}{galons} * \frac {1.609km}{miles} * \frac{1  galon}{3.785 l}$

### 0.425 differance
 

In [19]:
df.dtypes

mpg             object
cylinders       object
displacement    object
horsepower      object
weight          object
acceleration    object
model year      object
origin          object
name            object
dtype: object

In [ ]:
df["mpg"] = df["mpg"].astype(float) # changed coloumn type for converting its values 

In [22]:
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    mpg  cylinders  displacement  horsepower  weight  acceleration   
1     18          8           307         130    3504            12   
2     15          8           350         165    3693          11.5   
3     18          8           318         150    3436            11   
4     16          8           304         150    3433            12   
..   ...        ...           ...         ...     ...           ...   
394   27          4           140          86    2790          15.6   
395   44          4            97          52    2130          24.6   
396   32          4           135          84    2295          11.6   
397   28          4           120          79    2625          18.6   
398   31          4           119          82    2720          19.4   

     model year  origin                       name  
0    model year  origin                   car name  
1            70       1  chevrolet chevelle malibu  
2            70       1          buick skylark 320  
3            70       1         plymouth satellite  
4            70       1              amc rebel sst  
..          ...     ...                        ...  
394          82       1            ford mustang gl  
395          82       2                  vw pickup  
396          82       1              dodge rampage  
397          82       1                ford ranger  
398          82       1                 chevy s-10  

[399 rows x 9 columns]

In [8]:
df["kml"] = (df["mpg"] * 0.425).round(1) # rounded kml coloum as it is formated for the table auto-mpg

In [9]:
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0        130    3504          12.0   
1    15.0          8         350.0        165    3693          11.5   
2    18.0          8         318.0        150    3436          11.0   
3    16.0          8         304.0        150    3433          12.0   
4    17.0          8         302.0        140    3449          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0         86    2790          15.6   
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model year  origin                       name   kml  
0            70       1  chevrolet chevelle malibu   7.6  
1            70       1          buick skylark 320   6.4  
2            70       1         plymouth satellite   7.6  
3            70       1              amc rebel sst   6.8  
4            70       1                ford torino   7.2  
..          ...     ...                        ...   ...  
393          82       1            ford mustang gl  11.5  
394          82       2                  vw pickup  18.7  
395          82       1              dodge rampage  13.6  
396          82       1                ford ranger  11.9  
397          82       1                 chevy s-10  13.2  

[398 rows x 10 columns]

## $(L100km) = \frac{100}{kml}$

In [11]:
df["L_100km"] = (100 / df["kml"]).round(1)
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0        130    3504          12.0   
1    15.0          8         350.0        165    3693          11.5   
2    18.0          8         318.0        150    3436          11.0   
3    16.0          8         304.0        150    3433          12.0   
4    17.0          8         302.0        140    3449          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0         86    2790          15.6   
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model year  origin                       name   kml  L_100km  
0            70       1  chevrolet chevelle malibu   7.6     13.2  
1            70       1          buick skylark 320   6.4     15.6  
2            70       1         plymouth satellite   7.6     13.2  
3            70       1              amc rebel sst   6.8     14.7  
4            70       1                ford torino   7.2     13.9  
..          ...     ...                        ...   ...      ...  
393          82       1            ford mustang gl  11.5      8.7  
394          82       2                  vw pickup  18.7      5.3  
395          82       1              dodge rampage  13.6      7.4  
396          82       1                ford ranger  11.9      8.4  
397          82       1                 chevy s-10  13.2      7.6  

[398 rows x 11 columns]

In [14]:
df.sort_values("L_100km").head(5)[["name", "kml", "L_100km"]]


name   kml  L_100km
322             mazda glc  19.8      5.1
329   honda civic 1500 gl  19.0      5.3
394             vw pickup  18.7      5.3
325  vw rabbit c (diesel)  18.8      5.3
326    vw dasher (diesel)  18.4      5.4

## Answer:

In result, we can say that the most cheap car in terms of gas usuage is ***mazda glc*** with 5.1 liters per 100 km fule consumption